### PR TITLE
Relax ownership requirements in PiFac using new proof trait.

### LINKED ISSUE #402: Address ownership requirements in pifac
- [ ] Apply the solution in #135 to the `pifac` proof (see epic for details)

### PR DESCRIPTION
**Pull request by gatoWololo**
Closes #402

### CODE REVIEW

**Comment by gatoWololo** [comment]
@gatoWololo This is now ready to review!

File: src/zkp/pifac.rs [hunk 1]
```diff
@@ -329,136 +332,143 @@ fn sqrt(num: &BigNumber) -> BigNumber {
...
             )?;
-            let mut transcript = Transcript::new(b"PiFac Test");
+
             assert!(incorrect_factors
-                .verify(&input, &(), &mut transcript)
+                .verify(input, &(), &mut transcript())
                 .is_err());
 
-            let mut transcript = Transcript::new(b"PiFac Test");
             let small_p = BigNumber::from(7u64);
```
Comment (marsella) [inline src/zkp/pifac.rs:423]: Todo: add docs "factors must be in the valid range (e.g. large enough)."

File: src/zkp/pifac.rs [hunk 2]
```diff
@@ -329,136 +332,143 @@ fn sqrt(num: &BigNumber) -> BigNumber {
...
+
             assert!(small_proof
-                .verify(&small_input, &(), &mut transcript)
+                .verify(small_input, &(), &mut transcript())
                 .is_err());
 
-            let mut transcript = Transcript::new(b"PiFac Test");
-            let regular_sized_q = prime_gen::try_get_prime_from_pool_insecure(&mut rng).unwrap();
-            let mixed_input = CommonInput::new(&setup_params, &(&small_p * &regular_sized_q));
+            let regular_sized_q = prime_gen::try_get_prime_from_pool_insecure(&mut rng2).unwrap();
```
Comment (marsella) [inline src/zkp/pifac.rs:440]: Same other note as before: this should also test the case where exactly one of the factors is out of range. I guess we do it for small p but not small q...?

File: src/zkp/pifac.rs [hunk 3]
```diff
@@ -329,136 +332,143 @@ fn sqrt(num: &BigNumber) -> BigNumber {
...
             )?;
-            let mut transcript = Transcript::new(b"PiFac Test");
+
             assert!(mixed_proof
-                .verify(&mixed_input, &(), &mut transcript)
+                .verify(mixed_input, &(), &mut transcript())
                 .is_err());
 
-            let mut transcript = Transcript::new(b"PiFac Test");
             let small_fac_p = &not_p0 * &BigNumber::from(2u64);
```
Comment (marsella) [inline src/zkp/pifac.rs:455]: Todo: add docs here:" modulus cannot have small factors".

For this particular test, I would rather put it in its own function that does not call the with-random-small-factors-proof function because I think it does not use any of the fields provided by that.

**Review by marsella — CHANGES_REQUESTED** [review]
Thanks for this pr! Your changes overall look good; I think the necessary thing for this pr is to add some documentation for clarity.

More broadly I think these tests are kind of incomplete and could be organized more nicely. That can be handled in #113, I think it's worth adding a comment to that issue with the additional notes here.

File: src/zkp/pifac.rs [hunk 4]
```diff
@@ -329,135 +332,162 @@ fn sqrt(num: &BigNumber) -> BigNumber {
...
             let small_q = BigNumber::from(11u64);
-            let setup_params = VerifiedRingPedersen::gen(&mut rng, &())?;
-            let small_input = CommonInput::new(&setup_params, &(&small_p * &small_q));
+            let setup_params = VerifiedRingPedersen::gen(&mut rng2, &())?;
+            let modulus = &small_p * &small_q;
+            let small_input = CommonInput::new(&setup_params, &modulus);
             let small_proof = PiFacProof::prove(
-                &input,
-                &ProverSecret::new(&small_p, &small_q),
+                input,
```
Comment (marsella) [inline src/zkp/pifac.rs:435]: note: this should be `small_input`, not input; in this case we want to test that a proof with invalid inputs (e.g. small secrets) that is constructed and validated correctly will not verify.

Similarly, below, the `mixed_input` should be used for both proving and verifying.

I'm making this comment so it's easier to reference from #113; you don't need to fix it here because when I tested this change the test fails, so it's a pre-existing cryptography problem.

**Review by marsella — APPROVED** [review]
Thanks, this looks good!
